### ISSUE TITLE
num_classes no longer changing the classifier

### ISSUE THREAD

**Issue by ihamdi** [mlmed/torchxrayvision#76]
Today I noticed that all predictions by the "All" model were "16" despite the dataset only giving labels between 0 and 3. First I thought it was a mistake on my end but I knew that I didn't change anything today. I checked the code anyways and found out that setting num_classes to 4 still gives a model.classifier of Linear(in_features=1024, out_features=18, bias=True).

Using normal Densenet121 from torchvision models works fine so it is not an issue with my datamodule or data. The very strange part is that this changed all of a sudden while it was searching through hyperparameters with optuna. First 3 or 4 trials were fine and then all of a sudden I see label "4" and even "5" on the subsequent runs.

**Comment by ieee8023** [comment]
When loading the "all" model, `num_classes` shouldn't work because it would not be clear what classes to select. `num_classes` should work for a model which is not loaded from weights. I will add some unit tests to be sure. I also think an error should be raised if `num_classes` and `weights` are both specified. Does this design make sense?

**Comment by ihamdi** [comment]
> When loading the "all" model, `num_classes` shouldn't work because it would not be clear what classes to select. `num_classes` should work for a model which is not loaded from weights. I will add some unit tests to be sure. I also think an error should be raised if `num_classes` and `weights` are both specified. Does this design make sense?

It does make sense. However, I've been using that for the past month and a half on the [SIIM-FISABIO-RSNA COVID-19 Detection](https://www.kaggle.com/c/siim-covid19-detection) competition on Kaggle. The ALL model has been the best. But suddenly today, it started predicting everything to be label 16. Its very strange. I must have 20-30 runs on wandb using the same code. 

I used to pass num_classes = 2 when I was doing classification (by considering any label above 0 to be 1) and I was getting around 80% on validation accuracy. I started multiclass (num_classes = 4) last week and was getting just over 60% with certain hyperparameters.

I'm going to download my code from last week and try that just in case I changed something by mistake but then again this started happening within the same optuna sweep.

**Comment by ieee8023** [comment]
You could have passed `num_classes = 2` and during training those classes were always the highest if you were using an argmax over the outputs. But if the model wasn't trained well the other classes could higher.

**Comment by ihamdi** [comment]
> You could have passed `num_classes = 2` and during training those outputs were always the highest if you were using an argmax over the outputs. But if the model wasn't trained well the other outputs could higher.

I am using argmax. Its just extremely strange how it was never an issue before. My confusion matrix always showed 2 or 4 labels according to the num_classes. Now its showing 5 (0-4) and one time it was 6.

I just downloaded my last version from github and I'm getting the same issue. Now everything is given label 4.

**Comment by ieee8023** [comment]
The latest version should have caused your code to raise an error when you specify num_classes and weights right? I'll close this for now but feel free to comment again discuss another solution.
